src/Streamly/Internal/Data/Array/Mut/Type.hs

```diff
@@ -0,0 +1,535 @@
+    -> Int -- ^ length of the slice
+    -> Array a
+    -> Array a
+getSlice index len arr@(Array{..}) =
+    if index >= 0 && len >= 0 && index + len <= arrLen
+    then arr {arrStart = arrStart + index, arrLen = len}
+    else error
+             $ "getSlice: invalid slice, index "
+             ++ show index ++ " length " ++ show len
+
+-------------------------------------------------------------------------------
+-- to Lists and streams
+-------------------------------------------------------------------------------
+
+-- | Convert an 'Array' into a list.
+--
+-- /Pre-release/
+{-# INLINE toList #-}
+toList :: MonadIO m => Array a -> m [a]
+toList arr@(Array {..}) = mapM (getIndexUnsafe arr) [0 .. (arrLen - 1)]
+
+-- | Use the 'read' unfold instead.
+--
+-- @toStreamD = D.unfold read@
+--
+-- We can try this if the unfold has any performance issues.
+{-# INLINE_NORMAL toStreamD #-}
+toStreamD :: MonadIO m => Array a -> D.Stream m a
+toStreamD arr@(Array {..}) =
+    D.mapM (getIndexUnsafe arr) $ D.enumerateFromToIntegral 0 (arrLen - 1)
+
+{-# INLINE toStreamK #-}
+toStreamK :: MonadIO m => Array a -> K.Stream m a
+toStreamK arr@(Array {..}) = K.mapM (getIndexUnsafe arr) $ K.unfoldr step 0
+
+    where
+
+    step i
+        | i == arrLen = Nothing
+        | otherwise = Just (i, i + 1)
+
+
+-------------------------------------------------------------------------------
+-- Folds
+-------------------------------------------------------------------------------
+
+-- | Like 'writeN' but does not check the array bounds when writing. The fold
+-- driver must not call the step function more than 'n' times otherwise it will
+-- corrupt the memory and crash. This function exists mainly because any
+-- conditional in the step function blocks fusion causing 10x performance
+-- slowdown.
+--
+-- /Pre-release/
+{-# INLINE_NORMAL writeNUnsafe #-}
+writeNUnsafe :: MonadIO m => Int -> Fold m a (Array a)
+writeNUnsafe n = Fold step initial return
+
+    where
+
+    initial = FL.Partial <$> newArray (max n 0)
+
+    step arr x = FL.Partial <$> snocUnsafe arr x
+
+-- | @writeN n@ folds a maximum of @n@ elements from the input stream to an
+-- 'Array'.
+--
+-- >>> writeN n = Fold.take n (Array.writeNUnsafe n)
+--
+-- /Pre-release/
+{-# INLINE_NORMAL writeN #-}
+writeN :: MonadIO m => Int -> Fold m a (Array a)
+writeN n = FL.take n $ writeNUnsafe n
+
+-------------------------------------------------------------------------------
+-- Unfolds
+-------------------------------------------------------------------------------
+
+-- | Resumable unfold of an array.
+--
+{-# INLINE_NORMAL producer #-}
+producer :: MonadIO m => Producer m (Array a) a
+producer = Producer step return return
+
+    where
+
+    {-# INLINE_LATE step #-}
+    step Array {..}
+        | assert (arrLen >= 0) (arrLen == 0) = return D.Stop
+    step arr@(Array {..}) = do
+        x <- getIndexUnsafe arr 0
+        return $ D.Yield x $ arr {arrStart = arrStart + 1, arrLen = arrLen - 1}
```

Comment:
May be better to use a simpler custom data type for the state in this loop. Would be easier to fuse in more cases.

Reply:
Except one field all other fields are in use.

Reply:
Yes, I know, and sometimes that one field itself makes a difference.